src/Storages/System/StorageSystemReplicas.cpp

```diff
@@ -24,9 +28,165 @@ namespace CurrentMetrics
 namespace DB
 {
 
+/// Allows to "deduplicate" getStatus() requests for the same table: if a request for a table is already in progress
+/// then the new request will return the same future as the previous one.
+class StatusRequestsPool
+{
+public:
+    struct RequestInfo
+    {
+        UInt64 request_id = 0;
+        std::shared_future<ReplicatedTableStatus> future;
+    };
+
+private:
+    ThreadPool thread_pool;
+
+    std::mutex mutex;
+    /// All requests from the queries that are currently being executed.
+    std::unordered_map<StoragePtr, RequestInfo> current_requests TSA_GUARDED_BY(mutex);
+    /// Requests that were added by currently executing queries but have not been scheduled yet.
+    std::deque<std::tuple<UInt64, StoragePtr, std::shared_ptr<std::promise<ReplicatedTableStatus>>, bool>> requests_to_schedule TSA_GUARDED_BY(mutex);
```

Comment:
not sure how it will behave if we have a promise that is not set, maybe set them in destructor?

> if the shared state is not ready, stores an exception object of type [std::future_error](https://en.cppreference.com/w/cpp/thread/future_error) with an error condition [std::future_errc::broken_promise](https://en.cppreference.com/w/cpp/thread/future_errc), makes the shared state ready and releases it.